pkg/keyspace/tso_keyspace_group.go

```diff
@@ -81,38 +81,146 @@ func (m *GroupManager) GetKeyspaceGroupByID(id uint32) (*endpoint.KeyspaceGroup,
+func (m *GroupManager) updateKeyspaceGroups(keyspaceGroups []*endpoint.KeyspaceGroup) error {
+	return m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+		for _, keyspaceGroup := range keyspaceGroups {
+			// Check if the keyspace group exists.
+			kg, err := m.store.LoadKeyspaceGroup(txn, keyspaceGroup.ID)
+			if err != nil {
+				return err
+			}
+			if kg == nil {
+				return ErrKeyspaceGroupNotFound
+			}
+			// TODO: add replica count
+			m.store.SaveKeyspaceGroup(txn, &endpoint.KeyspaceGroup{
+				ID:        keyspaceGroup.ID,
+				UserKind:  keyspaceGroup.UserKind,
+				Members:   keyspaceGroup.Members,
+				Keyspaces: keyspaceGroup.Keyspaces,
+				InSplit:   keyspaceGroup.InSplit,
+			})
 		}
 		return nil
 	})
 }
 
-// GetAvailableKeyspaceGroupIDByKind returns the available keyspace group id by user kind.
+// GetAvailableKeyspaceGroupIDByKind returns the available keyspace group ID by user kind.
 func (m *GroupManager) GetAvailableKeyspaceGroupIDByKind(userKind endpoint.UserKind) (string, error) {
 	// TODO: implement it
 	return "0", nil
 }
+
+// SplitKeyspaceGroupByID splits the keyspace group by ID into a new keyspace group with the given new ID.
+// And the keyspaces in the old keyspace group will be moved to the new keyspace group.
+func (m *GroupManager) SplitKeyspaceGroupByID(id, newID uint32, keyspaces []uint32) error {
+	return m.store.RunInTxn(m.ctx, func(txn kv.Txn) error {
+		// Load the old keyspace group first.
+		oldKg, err := m.store.LoadKeyspaceGroup(txn, id)
+		if err != nil {
+			return err
+		}
+		if oldKg == nil {
+			return ErrKeyspaceGroupNotFound
+		}
+		// Check if the new keyspace group already exists.
+		newKg, err := m.store.LoadKeyspaceGroup(txn, newID)
+		if err != nil {
+			return err
+		}
+		if newKg != nil {
+			return ErrKeyspaceGroupExists
+		}
+		// Check if the keyspaces are all in the old keyspace group.
+		if len(keyspaces) > len(oldKg.Keyspaces) {
+			return ErrKeyspaceNotInKeyspaceGroup
+		}
+		var (
+			oldKeyspaceMap = make(map[uint32]struct{}, len(oldKg.Keyspaces))
+			newKeyspaceMap = make(map[uint32]struct{}, len(keyspaces))
+		)
+		for _, keyspace := range oldKg.Keyspaces {
+			oldKeyspaceMap[keyspace] = struct{}{}
+		}
+		for _, keyspace := range keyspaces {
+			if _, ok := oldKeyspaceMap[keyspace]; !ok {
+				return ErrKeyspaceNotInKeyspaceGroup
+			}
+			newKeyspaceMap[keyspace] = struct{}{}
+		}
+		// Get the split keyspace group for the old keyspace group.
+		splitKeyspaces := make([]uint32, 0, len(oldKg.Keyspaces)-len(keyspaces))
+		for _, keyspace := range oldKg.Keyspaces {
+			if _, ok := newKeyspaceMap[keyspace]; !ok {
+				splitKeyspaces = append(splitKeyspaces, keyspace)
+			}
+		}
+		// Update the old keyspace group.
+		oldKg.Keyspaces = splitKeyspaces
+		m.store.SaveKeyspaceGroup(txn, oldKg)
+		// Create the new split keyspace group.
+		return m.store.SaveKeyspaceGroup(txn, &endpoint.KeyspaceGroup{
```

Comment:
Though, the put events to the old and new groups will be propagated asynchronously, we should still update/create both keyspace groups in a transaction to provide atomic operation.

Reply:
The whole `SplitKeyspaceGroupByID` is in a txn.